File: cpp/baekjoon/baekjoon_1854.cpp
// baekjoon_1854 K번째 최단 경로 찾기

//v1
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#define INF 987654321

using namespace std;

vector<pair<int, int>> adj_list[1001];

vector<vector<int>> dijkstra(int K) {
	vector<vector<int>> visited(1001, vector<int>(K, INF));
	priority_queue < pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
	pq.push({ 0, 1 });
	visited[1][0] = 0;

	while (!pq.empty()) {
		int dist = pq.top().first;
		int cur_node = pq.top().second;
		pq.pop();

		for (int i = 0; i < adj_list[cur_node].size(); ++i) {
			int next_node = adj_list[cur_node][i].first;
			int next_dist = adj_list[cur_node][i].second;
			int cost = dist + next_dist;

			if (cost < visited[next_node][K - 1]) {
				pq.push({ cost, next_node });
				visited[next_node][K - 1] = cost;
				sort(visited[next_node].begin(), visited[next_node].end());
			}
		}


	}

	return visited;
}


int main() {
	int N, M, K;
	cin >> N >> M >> K;

	int a, b, c;
	for (int m = 0; m < M; ++m) {
		cin >> a >> b >> c;
		adj_list[a].push_back({ b, c });
	}

	vector<vector<int>> visited = dijkstra(K);

	for (int i = 1; i < N + 1; ++i) {
		if (visited[i][K - 1] == INF) cout << -1 << '\n';
		else cout << visited[i][K - 1] << '\n';

	}
}

File: python/baekjoon/baekjoon_1854.py
# baekjoon_1854 K번째 최단경로 찾기

# v1
# import sys, heapq, math
#
# input = sys.stdin.readline
# INF = math.inf
#
#
# def dijkstra():
#     hq = []
#     heapq.heappush(hq, [0, 1])
#     visited[1][K-1] = 0
#     visited[1].sort()
#
#     while hq:
#         dist, cur_node = heapq.heappop(hq)
#
#         for next_node, next_dist in adj_list[cur_node]:
#             cost = next_dist + dist
#             if cost < visited[next_node][K-1]:
#                 heapq.heappush(hq, [cost, next_node])
#                 visited[next_node][K-1] = cost
#                 visited[next_node].sort()
#
#
# N, M, K = map(int, input().split())
# visited = [[INF]*K for _ in range(N+1)]
# adj_list = [[] for _ in range(N+1)]
#
# for _ in range(M):
#     a, b, c = map(int, input().split())
#     adj_list[a].append([b, c])
#
# dijkstra()
#
# for i in range(1, N+1):
#     if visited[i][K-1] == INF:
#         print(-1)
#     else:
#         print(visited[i][K-1])


# v2
import sys, heapq, math

input = sys.stdin.readline
INF = math.inf


def dijkstra():
    hq = []
    heapq.heappush(hq, [0, 1])
    visited[1][0] = 0

    while hq:
        dist, cur_node = heapq.heappop(hq)

        for next_node, next_dist in adj_list[cur_node]:
            cost = next_dist + dist
            if cost < visited[next_node][K-1]:
                heapq.heappush(hq, [cost, next_node])
                visited[next_node][K-1] = cost
                visited[next_node].sort()


N, M, K = map(int, input().split())
visited = [[INF]*K for _ in range(N+1)]
adj_list = [[] for _ in range(N+1)]

for _ in range(M):
    a, b, c = map(int, input().split())
    adj_list[a].append([b, c])

dijkstra()

for i in range(1, N+1):
    if visited[i][K-1] == INF:
        print(-1)
    else:
        print(visited[i][K-1])


